assets/shared/helpers/player/use-editor-player.js

```diff
@@ -0,0 +1,168 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { store as coreStore } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import Player from './index';
+
+const API_SCRIPT_ID = 'player-script';
+const YOUTUBE_API_SRC = 'https://www.youtube.com/iframe_api';
+const VIMEO_API_SRC = 'https://player.vimeo.com/api/player.js';
+
+/**
+ * Hook to get the dependencies used as trigger to (re)set the player instance.
+ *
+ * @param {Object} videoBlock Video block object.
+ *
+ * @return {Object} Object containing embed fetching and block selected state.
+ */
+const useTriggerDependencies = ( videoBlock ) => {
+	// Check embed fetching.
+	const { fetching } = useSelect(
+		( select ) => ( {
+			fetching: select( coreStore ).isRequestingEmbedPreview(
+				videoBlock?.attributes?.url
+			),
+		} ),
+		[ videoBlock?.attributes?.url ]
+	);
+
+	// Check if block is selected. We need to get the player reference again when it's video block
+	// because it re-creates the video element when it's (un)selected.
+	const { isBlockSelected } = useSelect(
+		( select ) => ( {
+			isBlockSelected: select( blockEditorStore ).isBlockSelected(
+				videoBlock.clientId
+			),
+		} ),
+		[ videoBlock.clientId ]
+	);
+
+	return { fetching, isBlockSelected };
+};
+
+/**
+ * A wrapper to useEffect with a timeout of 1 milisecond, in order to delay the sync execution.
+ *
+ * @param {Function} effect Effect callback.
+ * @param {Array}    deps   Effect dependencies.
+ */
+const useDelayedEffect = ( effect, deps ) => {
+	useEffect( () => {
+		setTimeout( () => {
+			effect();
+		}, 1 );
+	}, deps ); // eslint-disable-line react-hooks/exhaustive-deps -- Wrapper to useEffect.
+};
+
+/**
+ * Add a script to a body.
+ *
+ * @param {HTMLBodyElement} body   The body where the script will be appended.
+ * @param {string}          src    Script src.
+ * @param {Function}        onLoad Script load callback.
+ */
+const addScript = ( body, src, onLoad ) => {
```

Comment:
TBH I'm skeptical about adding these deps from JS. Is it a common practice? I realize that we don't really want to include it unconditionally from the server-side, and we potentially need to add it dynamically when the right blocks are added. So probably it's fine, but it just seems a bit strange to me...

Reply:
The main point here is that we need to add them [inside the sandbox](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/embed/embed-preview.js#L100) (an iframe that renders the video iframe).

I started trying a different approach, adding the scripts to the editor normally, and instance the players from the iframes, but it didn't work with one of the APIs that I started testing (IIRC, it was YouTube). So I needed to add the scripts inside the sandbox, and I think it would be the only way to do that.

Reply:
Oh interesting. That makes sense then 👍